parlai/core/torch_generator_agent.py

```diff
@@ -705,7 +708,13 @@ def compute_loss(self, batch, return_output=False):
         """
         if batch.label_vec is None:
             raise ValueError('Cannot compute loss without a label.')
-        model_output = self.model(*self._model_input(batch), ys=batch.label_vec)
+        r3f_adjustment = None
```

Comment:
Now that we've got the context manager, I think we can implement this as a mixin pretty well, approximately:

```python
class R3FMixin:
    def compute_loss(self, batch, return_output)
          loss, standard_output = super().compute_loss(batch, return_output=True)
          standard_scores, *_ = standard_output
          with R3FNoiseEmbeddingContextManager(self.model) as r3f:
              perturbed_output = self.model(*self._model_input(batch), ys=batch.label_vec)
              perturbed_scores, *_ = perturbed_output
              r3floss = r3f.compute_sym_loss(standard_scores, perturbed_scores)
              loss += r3floss[notnull].mean()
         if return_output:
              return loss, standard_output
         else:
              return loss

class R3FTransformerGeneratorAgent(R3FMixin, TransformerGeneratorAgent):
      pass
```

I've got in my mind a possible way that we could have losses be more generically implemented (basically migrate us to how fairseq/classyvision do it).

Reply:
Thinking about it a little more, I'm not actually quite sure this works in the context of BART.

Specifically, BART overrides `compute_loss` with its own loss function to remove the start token. Implementing R3F as a Mixin makes the `super()` line in the code above sort of wonky.

Reply:
(In general, I do think "composition over inheritance" might be a decent pattern to go for here, but we can chat offline in more depth. I'm especially curious about how fairseq/classyvision currently has it done, if you have code pointers.)

Reply:
Dope, would love to talk about how we can make this a composition.

Reply:
Update: Chatted offline with @stephenroller about what the inheritance tree here would look like. 

Doing 

```
class BartAgent_R3F(R3FMixin, BartAgent):
    pass
```

so that's the final class works.